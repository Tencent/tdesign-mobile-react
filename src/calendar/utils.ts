import { TCalendarValue } from './type';

export function getMonthByOffset(date: TCalendarValue, offset: number) {
  const innerDate = new Date(date);
  innerDate.setMonth(innerDate.getMonth() + offset);
  innerDate.setDate(1);
  return innerDate;
}

export function getYearByOffset(date: TCalendarValue, offset: number) {
  const innerDate = new Date(date);
  innerDate.setFullYear(innerDate.getFullYear() + offset);
  innerDate.setDate(1);
  return innerDate;
}

export const getPrevMonth = (date: TCalendarValue) => getMonthByOffset(date, -1);
export const getNextMonth = (date: TCalendarValue) => getMonthByOffset(date, 1);
export const getPrevYear = (date: TCalendarValue) => getYearByOffset(date, -1);
export const getNextYear = (date: TCalendarValue) => getYearByOffset(date, 1);
