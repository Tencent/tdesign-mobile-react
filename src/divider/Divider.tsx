import React from 'react';
import type { FC } from 'react';
import classNames from 'classnames';
import { TdDividerProps } from './type';
import { dividerDefaultProps } from './defaultProps';
import { StyledProps } from '../common';
import useConfig from '../hooks/useConfig';
import useDefaultProps from '../hooks/useDefaultProps';
import parseTNode from '../_util/parseTNode';

export interface DividerProps extends TdDividerProps, StyledProps {}

const Divider: FC<DividerProps> = (props) => {
  const { children, align, dashed, layout, content, style } = useDefaultProps<DividerProps>(props, dividerDefaultProps);
  const { classPrefix } = useConfig();
  const dividerClass = `${classPrefix}-divider`;
  const contentNode = content || children;

  const classes = classNames([
    dividerClass,
    `${dividerClass}--${layout}`,
    `${dividerClass}--${align}`,
    {
      [`${dividerClass}--dashed`]: dashed,
    },
  ]);

  return (
    <div className={classes} style={style} role="separator">
      <div className={`${dividerClass}__content`}>{parseTNode(contentNode)}</div>
    </div>
  );
};

Divider.displayName = 'Divider';

export default Divider;
