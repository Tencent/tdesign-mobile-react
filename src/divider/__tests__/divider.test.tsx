import React from 'react';
import { describe, expect, it, render } from '@test/utils';
import Divider from '../Divider';

describe('Divider', () => {
  describe('props', () => {
    it(':align', () => {
      const aligns = ['left', 'right', 'center'] as const;

      function checkAlign(align: (typeof aligns)[number]) {
        const { container } = render(<Divider align={align}>Text</Divider>);
        expect(container.firstChild).toHaveClass(`t-divider--${align}`);
      }

      aligns.forEach(checkAlign);
    });

    it(':layout', () => {
      const layouts = ['horizontal', 'vertical'] as const;

      function checkLayout(layout: (typeof layouts)[number]) {
        const { container } = render(<Divider layout={layout} />);
        expect(container.firstChild).toHaveClass(`t-divider--${layout}`);
      }

      layouts.forEach(checkLayout);
    });

    it(':content', () => {
      // Test content as a function
      const { container: container1 } = render(<Divider content={() => <span className="custom-node">TNode</span>} />);
      expect(container1.querySelector('.custom-node')).toBeTruthy();
      expect(container1).toMatchSnapshot();

      // Test content as a string
      const { container: container2 } = render(<Divider content="TDesign" />);
      expect(container2.textContent).toBe('TDesign');
      expect(container2).toMatchSnapshot();
    });

    it(':dashed', () => {
      // Test default value (false)
      const { container: container1 } = render(<Divider />);
      expect(container1.firstChild).not.toHaveClass('t-divider--dashed');

      // Test dashed = true
      const { container: container2 } = render(<Divider dashed={true} />);
      expect(container2.firstChild).toHaveClass('t-divider--dashed');

      // Test dashed = false
      const { container: container3 } = render(<Divider dashed={false} />);
      expect(container3.firstChild).not.toHaveClass('t-divider--dashed');
    });
  });

  describe('slot', () => {
    it(':default', () => {
      const { container } = render(<Divider>TDesign</Divider>);
      expect(container).toMatchSnapshot();
    });

    it(':content', () => {
      const testSlotContentId = 'divider-content-test-slot-id';
      const { container } = render(
        <Divider>
          <span className="custom-node" data-testid={testSlotContentId}>
            TNode
          </span>
        </Divider>,
      );
      expect(container.querySelector('.custom-node')).toBeTruthy();
      expect(container.querySelector(`[data-testid="${testSlotContentId}"]`)).toBeTruthy();
      expect(container).toMatchSnapshot();
    });
  });

  it('content prop and default slot exist meanwhile', () => {
    const { container } = render(<Divider content="prop content">TDesign</Divider>);
    expect(container.textContent).toBe('prop content');
  });

  describe('vertical divider', () => {
    it('applies dashed class on vertical divider', () => {
      const { container } = render(<Divider layout="vertical" dashed={true} />);
      expect(container.firstChild).toHaveClass('t-divider--dashed');
      expect(container).toMatchSnapshot();
    });

    it('applies align class on vertical divider', () => {
      const { container } = render(
        <Divider layout="vertical" align="left">
          text
        </Divider>,
      );
      expect(container.firstChild).toHaveClass('t-divider--left');
      expect(container).toMatchSnapshot();
    });
  });
});
