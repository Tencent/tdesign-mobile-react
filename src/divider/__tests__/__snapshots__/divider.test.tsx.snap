// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Divider > props > :content 1`] = `
<div>
  <div
    class="t-divider t-divider--horizontal t-divider--center"
    role="separator"
  >
    <div
      class="t-divider__content"
    >
      <span
        class="custom-node"
      >
        TNode
      </span>
    </div>
  </div>
</div>
`;

exports[`Divider > props > :content 2`] = `
<div>
  <div
    class="t-divider t-divider--horizontal t-divider--center"
    role="separator"
  >
    <div
      class="t-divider__content"
    >
      TDesign
    </div>
  </div>
</div>
`;

exports[`Divider > slot > :content 1`] = `
<div>
  <div
    class="t-divider t-divider--horizontal t-divider--center"
    role="separator"
  >
    <div
      class="t-divider__content"
    >
      <span
        class="custom-node"
        data-testid="divider-content-test-slot-id"
      >
        TNode
      </span>
    </div>
  </div>
</div>
`;

exports[`Divider > slot > :default 1`] = `
<div>
  <div
    class="t-divider t-divider--horizontal t-divider--center"
    role="separator"
  >
    <div
      class="t-divider__content"
    >
      TDesign
    </div>
  </div>
</div>
`;

exports[`Divider > vertical divider > applies align class on vertical divider 1`] = `
<div>
  <div
    class="t-divider t-divider--vertical t-divider--left"
    role="separator"
  >
    <div
      class="t-divider__content"
    >
      text
    </div>
  </div>
</div>
`;

exports[`Divider > vertical divider > applies dashed class on vertical divider 1`] = `
<div>
  <div
    class="t-divider t-divider--vertical t-divider--center t-divider--dashed"
    role="separator"
  >
    <div
      class="t-divider__content"
    />
  </div>
</div>
`;
